src/V3Randomize.cpp

```diff
@@ -0,0 +1,280 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Expression width calculations
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2020 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3Randomize's Transformations:
+//
+// Each randomize() method call:
+//      Mark class of object on which randomize() is called
+// Mark all classes that inherit from previously marked classed
+// Mark all classes whose instances are randomized member variables of marked classes
+// Each marked class:
+//      define a virtual randomize() method that randomizes its random variables
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Randomize.h"
+
+//######################################################################
+// Visitor that marks classes needing a randomize() method
+
+class RandomizeMarkVisitor final : public AstNVisitor {
+private:
+    // NODE STATE
+    // Cleared on Netlist
+    //  AstClass::user1()       -> bool.  Set true to indicate needs randomize processing
+    AstUser1InUse m_inuser1;
+
+    typedef std::unordered_set<AstClass*> DerivedSet;
+    typedef std::unordered_map<AstClass*, DerivedSet> BaseToDerivedMap;
+
+    BaseToDerivedMap m_baseToDerivedMap;  // Mapping from base classes to classes that extend them
+
+    // METHODS
+    VL_DEBUG_FUNC;
+
+    void markMembers(AstClass* nodep) {
+        for (auto* classp = nodep; classp;
+             classp = classp->extendsp() ? classp->extendsp()->classp() : nullptr) {
+            for (auto* memberp = classp->stmtsp(); memberp; memberp = memberp->nextp()) {
+                // If member is rand and of class type, mark its class
+                if (VN_IS(memberp, Var) && VN_CAST(memberp, Var)->isRand()) {
+                    if (auto* classRefp = VN_CAST(memberp->dtypep(), ClassRefDType)) {
+                        auto* classp = classRefp->classp();
+                        markMembers(classp);
+                        markDerived(classp);
+                        classRefp->classp()->user1(true);
+                    }
+                }
+            }
+        }
+    }
+    void markDerived(AstClass* nodep) {
+        if (m_baseToDerivedMap.find(nodep) != m_baseToDerivedMap.end()) {
+            for (auto* classp : m_baseToDerivedMap.at(nodep)) {
```

Comment:
(approx)  this is faster.

```suggestion
        const auto it = m_baseToDerivedMap.find(nodep);
        if (it != m_baseToDerivedMap.end()) {
            for (auto* classp : *it) {
```